sdks/java/io/mongodb/src/main/java/org/apache/beam/sdk/io/mongodb/MongoDbIO.java

```diff
@@ -0,0 +1,535 @@
+        boolean isPreviousTwo = true;
+        for (Integer range : idRanges.values()) {
+          if (range == 2) {
+            splitId = splitIDFromMongo.get(indexPosition + 1);
+          } else {
+            if (isPreviousTwo) {
+              indexPosition++;
+              isPreviousTwo = false;
+            }
+            splitId = splitIDFromMongo.get(indexPosition);
+          }
+          indexPosition++;
+          splitIDCalculated.add(splitId);
+        }
+        return createSourceList(splitIDCalculated);
+      }
+      return createSourceList(splitIDFromMongo);
+    }
+
+    private List<BoundedSource> createSourceList(ArrayList splitIDList) {
+      List<BoundedSource> splitSourceList = new ArrayList<>();
+      String lastID = null; // Lower boundary of the first min split
+      int listIndex = 0;
+      for (final Object splitIdValue : splitIDList) {
+//          final Document  currentDoc = (Document) aSplitData;
+        String currentID = splitIdValue.toString();
+        String newFilter;
+        if (filter != null && !filter.isEmpty()) {
+          if (listIndex == 0) {
+            newFilter = "{ $and: [ {\"_id\":{$lte:ObjectId(\""
+                + currentID + "\")}}, "
+                + filter + " ]}";
+          } else if (listIndex == (splitIDList.size() - 1)) {
+            newFilter = "{ $and: [ {\"_id\":{$gt:ObjectId(\"" + lastID
+                + "\")," + "$lt:ObjectId(\"" + currentID
+                + "\")}}, " + filter + " ]}";
+            splitSourceList.add(new BoundedMongoDbSource
+                (uri, database, collection, newFilter, numberSplit));
+            newFilter = "{ $and: [ {\"_id\":{$gt:ObjectId(\"" + currentID
+                + "\")}}, " + filter + " ]}";
+          } else {
+            newFilter = "{ $and: [ {\"_id\":{$gt:ObjectId(\"" + lastID
+                + "\")," + "$lte:ObjectId(\"" + currentID + "\")}}, "
+                + filter + " ]}";
+          }
+        } else {
+          if (listIndex == 0) {
+            newFilter = "{\"_id\":{$lte:ObjectId(\"" + currentID + "\")}}";
+          } else if (listIndex == (splitIDList.size() - 1)) {
+            newFilter = "{\"_id\":{$gt:ObjectId(\"" + lastID + "\"),"
+                + "$lt:ObjectId(\"" + currentID + "\")}}";
+            splitSourceList.add(new BoundedMongoDbSource
+                (uri, database, collection, newFilter, numberSplit));
+            newFilter = "{\"_id\":{$gt:ObjectId(\"" + currentID + "\")}}";
+          } else {
+            newFilter = "{\"_id\":{$gt:ObjectId(\"" + lastID + "\"),"
+                + "$lte:ObjectId(\"" + currentID + "\")}}";
+          }
+        }
+        splitSourceList.add(new BoundedMongoDbSource
+            (uri, database, collection, newFilter, numberSplit));
+        lastID = currentID;
+        listIndex++;
+      }
+      return splitSourceList;
+    }
+  }
+
+  private static class BoundedMongoDbReader extends BoundedSource.BoundedReader {
+
+    private String uri;
```

Comment:
Rather than listing the configuration variables, can you store a single "private final BoundedMongoDbSource source"? That will be fewer code and easier to maintain.